src/SelfServe/PropertyDecorators.tsx

```diff
@@ -0,0 +1,108 @@
+import { DropdownItem, Info, InputType, UiType } from "../Explorer/Controls/SmartUi/SmartUiComponent";
+import { addPropertyToMap } from "./SelfServeUtils";
+
+interface Decorator {
+  name: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  value: any;
+}
+
+interface InputOptionsBase {
+  label: string;
+}
+
+export interface NumberInputOptions extends InputOptionsBase {
+  min: (() => Promise<number>) | number;
+  max: (() => Promise<number>) | number;
+  step: (() => Promise<number>) | number;
+  uiType: UiType;
+}
+
+export interface StringInputOptions extends InputOptionsBase {
+  placeholder?: (() => Promise<string>) | string;
+}
+
+export interface BooleanInputOptions extends InputOptionsBase {
+  trueLabel: (() => Promise<string>) | string;
+  falseLabel: (() => Promise<string>) | string;
+}
+
+export interface DropdownInputOptions extends InputOptionsBase {
+  choices: (() => Promise<DropdownItem[]>) | DropdownItem[];
+}
+
+type InputOptions = NumberInputOptions | StringInputOptions | BooleanInputOptions | DropdownInputOptions;
+
+function isNumberInputOptions(inputOptions: InputOptions): inputOptions is NumberInputOptions {
+  return !!(inputOptions as NumberInputOptions).min;
+}
+
+function isBooleanInputOptions(inputOptions: InputOptions): inputOptions is BooleanInputOptions {
+  return !!(inputOptions as BooleanInputOptions).trueLabel;
+}
+
+function isDropdownInputOptions(inputOptions: InputOptions): inputOptions is DropdownInputOptions {
+  return !!(inputOptions as DropdownInputOptions).choices;
+}
+
```

Comment:
JavaScript has syntax to do all of these without a cast:

`return "min" in inputOptions`